### ISSUE TITLE
Problem with monotonic longitude requirements

### ISSUE THREAD

**Issue by wreckdump** [tobac-project/tobac#382]
I would like to use tobac with the input that has non-monotonic longitude. For example, below attached picture shows the entire tropics (25S - 25N) with the longitude ranging from 30E - 27.75E, which means that in the middle of the Pacific Ocean the longitude's coordinate transitions from 180E to 180W, and so on. This makes the longitude non-monotonic. 

![scrshot 2023 12 06 14 40 45](https://github.com/tobac-project/tobac/assets/26279723/e7c0e940-fe35-40ea-93d3-c9361ac6c9fb)

How can I work around this issue?

**Comment by wreckdump** [comment]
Aha! Thank you so much! Just so happens to be that I am using version 1.5.1! I'll try this! Thank you again!

**Comment by w-k-jones** [comment]
To add to what @freemansw1 says, most of tobac will run fine with non-monotonic coordinates, however area calculation will fail and the latitude/longitude of features that cross the date line will be wrong due to using linear interpolation (tracking/segmentation will still work correctly because these use array index rather than lat/lon).

An easy fix to these issues would be to load your data in xarray and roll the data so that the longitude is monotonic, then convert to iris for use with tobac. e.g.:

```
import xarray as xr
bt = xr.DataArray("path/to/data.nc")
bt = bt.roll({"lon":offset}, roll_coords=True)
bt = bt.load().to_iris()

features = tobac.feature_detection(bt, ...)
```

The required offset is in the index values, so if your data has 0.1 degree grid spacing you would need to roll by 2100 to get the lon array to span -180->180.

Note also that PBCs don't work with predictive tracking in v1.5.1, but this is fixed for v1.5.2 which will be released very soon

**Comment by wreckdump** [comment]
Errm. I don't think the periodic boundary condition function is working properly. I was just testing out a sample data that I have made (https://file.io/XjmoIGLdXurn) with the following info.

```
dimensions:
	time = UNLIMITED ; // (22 currently)
	lat = 201 ;
	lon = 1440 ;
variables:
	float time(time) ;
		time:standard_name = "time" ;
		time:units = "hours since 2020-01-21 00:00:00" ;
	float lat(lat) ;
		lat:standard_name = "latitude" ;
		lat:units = "degree_north" ;
	float lon(lon) ;
		lon:standard_name = "longitude" ;
		lon:units = "degree_east" ;
	float tb(time, lat, lon) ;
		tb:_FillValue = 99999.f ;
		tb:standard_name = "brightness temperature" ;
		tb:units = "K" ;
```
And some of the snapshots of the file looks like below.

![scrshot 2023 12 06 17 21 17](https://github.com/tobac-project/tobac/assets/26279723/3a6beda0-e577-4fd2-a932-bb4ed5e5da8b)
![scrshot 2023 12 06 17 21 35](https://github.com/tobac-project/tobac/assets/26279723/07a66f96-c729-4a14-a5ae-fbe49fd6cddc)
![scrshot 2023 12 06 17 21 51](https://github.com/tobac-project/tobac/assets/26279723/d321dccc-296a-4078-9a7e-00377274a78e)

And tracked results are shown like below.

![trk pnc err](https://github.com/tobac-project/tobac/assets/26279723/bdc70e58-ec07-4a6c-b26a-c4ecaae5d42e)

I expected the track to be the same trajectory but it is separated in two. Here is the options that I have used.

```
   # feature detection
   param_fet={}
   param_fet['target']='minimum'
   param_fet['threshold']=[thres]
   param_fet['n_min_threshold']=4
   param_fet['position_threshold']='center'
   param_fet['sigma_threshold']=0.5
   param_fet['PBC_flag'] = 'both'
   
     
   # segmentation
   param_seg={}
   param_seg['target']='minimum'
   param_seg['method']='watershed'
   param_seg['threshold']=thres
   
   
  # linking 
   param_lnk={}
   param_lnk['v_max']=None
   param_lnk['d_max']=111.0e3*1 # 10 degrees in meters
   param_lnk['stubs']=1
   param_lnk['order']=2
   param_lnk['extrapolate']=0
   param_lnk['memory']=0
   param_lnk['adaptive_stop']=0.5
   param_lnk['adaptive_step']=0.95
   param_lnk['subnetwork_size']=10
   param_lnk['method_linking']= 'random'

```

I tried both `random` and `predict` for `method_linking`.

**Comment by w-k-jones** [comment]
> Errm. I don't think the periodic boundary condition function is working properly. I was just testing out a sample data that I have made (https://file.io/XjmoIGLdXurn) with the following info.

I think you can fix this by adding the `PBC_flag` keyword to segmentation and tracking, and also addting `min_h2` and `max_h2` for tracking as below:

> ```
>    # feature detection
>    param_fet={}
>    param_fet['target']='minimum'
>    param_fet['threshold']=[thres]
>    param_fet['n_min_threshold']=4
>    param_fet['position_threshold']='center'
>    param_fet['sigma_threshold']=0.5
>    param_fet['PBC_flag'] = 'hdim_2'
>    
>      
>    # segmentation
>    param_seg={}
>    param_seg['target']='minimum'
>    param_seg['method']='watershed'
>    param_seg['threshold']=thres
>    param_seg['PBC_flag'] = 'hdim_2'
>    
>    
>   # linking 
>    param_lnk={}
>    param_lnk['v_max']=None
>    param_lnk['d_max']=111.0e3*1 # 10 degrees in meters
>    param_lnk['stubs']=1
>    param_lnk['order']=2
>    param_lnk['extrapolate']=0
>    param_lnk['memory']=0
>    param_lnk['adaptive_stop']=0.5
>    param_lnk['adaptive_step']=0.95
>    param_lnk['subnetwork_size']=10
>    param_lnk['method_linking']= 'random'
>    param_lnk['PBC_flag'] = 'hdim_2'
>    param_lnk['min_h2'] = 0
>    param_lnk['max_h2'] = 1440
> ```
> 

Note that `min_h2` and `max_h2` specify the size of your input array in the PBC dimension

> I tried both `random` and `predict` for `method_linking`.

If you want to try PBCs with tracking right away, you could checkout the `RC_v1.5.x` branch of the repository, although it should be released on conda soon

This is useful insight on using PBCs, it would be nice to have a way of storing the PBC settings so that you don't have to repeat it...

**Comment by w-k-jones** [comment]
Are you running this in a jupyter notebook? I've encountered a similar bug this evening that running PBCs with numba which occurs in notebooks but not in the python interpreter

**Comment by wreckdump** [comment]
No. I am not using jupyter notebook. I am just using python 3.11.6 with all the modules installed via pip.

**Comment by wreckdump** [comment]
Number version that I am using is 0.58.1

**Comment by w-k-jones** [comment]
I am closing this issue as the PBC/numba related bug was resolved in #384. I'd also like to look into storing more information about the input data and boundary conditions in future versions of tobac, but this can wait until after the xarray transition.
